src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -3359,48 +3421,55 @@ private bool HasTopLevelNullabilityConversion(TypeSymbolWithAnnotations source,
                     }
                     break;
 
+                case ConversionKind.ImplicitThrow:
+                    break;
+
                 case ConversionKind.Unboxing:
-                    if (!operandType.IsNull && targetType.IsTypeParameter())
+                    if (!operandType.IsNull)
                     {
-                        resultAnnotation = operandType.GetValueNullableAnnotation();
+                        if (targetType.IsTypeParameter())
+                        {
+                            resultAnnotation = operandType.GetValueNullableAnnotation();
 
-                        if (resultAnnotation == NullableAnnotation.NotAnnotated)
+                            if (resultAnnotation == NullableAnnotation.NotAnnotated)
+                            {
+                                resultAnnotation = NullableAnnotation.NotNullable;
+                            }
+                        }
+                        else if (targetType.IsValueType)
                         {
-                            resultAnnotation = NullableAnnotation.NotNullable;
+                            resultAnnotation = (targetType.IsNullableType() && operandType.NullableAnnotation.IsAnyNullable()) ? NullableAnnotation.Nullable : NullableAnnotation.NotNullable;
```

Comment:
>operandType.NullableAnnotation.IsAnyNullable() [](start = 79, length = 46)

Should we specially handle operand type which is a type parameter constrained to specific value type or a nullable value type (this can happen with overriding), can we get here for a scenario like this?

Reply:
Are there scenarios where the compiler treats the type parameters in the overridden methods as convertible to the constrained types?
```c#
abstract class A<T>
{
    internal abstract void F<U>(object x) where U : T;
}
class B1 : A<int>
{
    internal override void F<U>(object x)
    {
        int y = (U)x; // CS0029: Cannot implicitly convert type 'U' to 'int'
    }
}
class B2 : A<int?>
{
    internal override void F<U>(object x)
    {
        _ = ((U)x).Value; // CS1061: 'U' does not contain a definition for 'Value'
    }
}
```

---
In reply to: [239936037](https://github.com/dotnet/roslyn/pull/31499#discussion_r239936037) [](ancestors = 239936037)